compiler/rustc_mir/src/transform/check_consts/ops.rs

```diff
@@ -208,9 +208,36 @@ impl NonConstOp for LiveDrop {
     }
 }
 
+#[derive(Debug)]
+pub struct CellBorrowBehindRef;
+impl NonConstOp for CellBorrowBehindRef {
+    fn status_in_item(&self, _: &ConstCx<'_, '_>) -> Status {
+        Status::Unstable(sym::const_refs_to_cell)
+    }
+    fn build_error(&self, ccx: &ConstCx<'_, 'tcx>, span: Span) -> DiagnosticBuilder<'tcx> {
+        feature_err(
+            &ccx.tcx.sess.parse_sess,
+            sym::const_refs_to_cell,
+            span,
+            "cannot borrow here, since the borrowed element may contain interior mutability",
+        )
+    }
+}
+
 #[derive(Debug)]
 pub struct CellBorrow;
 impl NonConstOp for CellBorrow {
+    fn status_in_item(&self, ccx: &ConstCx<'_, '_>) -> Status {
+        match ccx.const_kind() {
+            // The borrow checker does a much better job at handling these than we do
```

Comment:
Please add a reference to the discussion that justifies this -- in particular, some kind of FIXME noting that this is *not compatible* with having heap allocations in consts.

Reply:
Also I find the control flow here extremely confusing. I think the `hir::ConstContext::ConstFn` check should be in validation.rs, not here.

Reply:
There are other `Qualif`s in this file where we do the check like this. The only reason for this function to take the `ccx` is to allow us to make this distinction.

Reply:
Yeah the control flow is already hard to follow without your PR, that's right.^^

Reply:
Checking the `const_kind` in `ops.rs` makes sense when the *stabilization status* is different in different contexts.

But it makes no sense when the fundamental soundness consideration is different, as in this case. All soundness considerations should be in the other file. Are there existing checks in this file that are about soundness, not just stabilization?

Reply:
The only other one that is about soundness is `StaticAccess`, which forbids accesses to statics except in other statics.

Reply:
Yeah IMO that check should be in `validation.rs`.

Reply:
done